### DISCUSSION TITLE
Do not use Byte-Order-Marks in UTF-8, they are erroneous

### DISCUSSION THREAD

**Discussion by Sturwandan** [pascalabcnet/pascalabcnet#2492] [Ideas]
The Byte-Order-Mark is a pseudo-character, which is used by some applications to detect the byte order of UTF-16 and UTF-32 encodings. The confusion here strives from the fact that USC-2 (predecessor of UTF-16) and UTF-32 are simply arrays of integers (unsigned 16 bit and 32 bit correspondingly), holding Unicode character codes. And this means, that on Little-Endian machines such as Intel x86, character U+044B "ы" (Cyrillic letter "Yery") is represented as bytes 4B 04 and on Big-Endian systems it would be written 04 4B, while character U+4B04 "䬄" (Chinese character "Xù"), would be written as 04 4B on LE machine and 4B 04 on BE machine. 

So a character U+FEFF (former zero-width space) was chosen to be put in front of USC-2, UTF-16 or UTF-32 data to avoid confusion, because its byte-swapped cousin U+FFFE  ￾ is not a character, so it cannot be found in valid Unicode.

Unlike dumb, but simple formats from the above, UTF-8 uses a variable-length encoding, with the first byte always holding  the most significant bits of the Unicode character code, and then, depending on its other bits, the byte-length of complete character is determined. For example, Unicode characters below 128 are encoded in one byte of format 0xxx xxxx (binary), which means that UTF-8 is fully compatible with ASCII, while values of 128 and above, up to 2047 are encoded as 110x xxxx 10xx xxxx representation, where the first byte is starting with 110, and two 1's mean that complete character is two bytes long, and the remaining bytes always start with 10 and hold six bits of actual character values, which allows to represent character up to U+07FF in 2-byte UTF-8 sequences.
Three byte sequences have similar structure: 1110 xxxx 10xx xxxx 10xx xxxx and so on. The number of 1's followed by 0s tells us how long the code is. x-bits represent actual bits of the character code.

As you can see, UTF-8 is a well-defined format unlike UTF-16 and more than that, the first byte of a sequence can be easily told apart from latter bytes and well-formed UTF-8 is easy to distinguish from malformed one fully automatically, making any special marks unnecessary.

Since the main feature of UTF-8, compared to UTF-16, is full compatibility with ASCII, this means that BOM is not only unnecessary, but also harmful.

Nevertheless, some text editors erroneously insert BOM into UTF-8, where it makes no sense, and that, unfortunately, includes PascalABC.net IDE. I have tried to create [a bug report in there](https://github.com/pascalabcnet/pascalabcnetide/issues/209), but it was closed without the issue being fixed.

So do I need to reopen the bug report there or something? How do I get the bug to be fixed?

**Comment by spectatorBH** [comment]
Historically, outside of US, due to extended character set requirements, regional encoding standards have always been a much more compicated story than you plotted, especially under DOS/Windows: OEM/ANSI/WIN/ISO-standards, not counting specific IBM, MAC, UNIX and various other either non-official or purely local legacy encodings (e.g. KOI-8 RU/UA, GB or Big5 for Chinese, etc.). So, in international practice (at least from 80s) a 7-bit ASCII charset has almost always been just a common subset of other popular 8-bit encodings on PCs. That was long before any universal and internationally accepted multibyte standards such as MBCS/UCS/UTF became a reality and saw a wide adoption.

Since PascalABC.NET is an advanced reincarnation of its older sibling PascalABC (now defunct) **designed primarily for russian schools and russian-speaking students**, for source files it natively supports only two: an older popular win-1251 russian-compatible 8-bit encoding and contemporary universal UTF-8. In order to easily tell apart one from another upon opening, it uses a BOM-mark in a non-standard way (just as most other editors you mention do, I think). 

So, it's not a bug here, it's an intentional feature, although non-documented: *if a file has a BOM-mark, it opens as UTF-8, otherwise it opens as win-1251*. All new files created with IDE editor are always **UTF-8 + BOM + CR/LF** line breaks. Since win-1251 single-byte encoding is still widely used in russian-oriented environment, it's impossible for them to drop it in favour of pure UTF-8 without BOM -- statistical analysis of file content is not a reliable method to recognize proper encoding when you might have very few russian chars inside a source code.

Yes, there're also some issues with keeping non-Windows line breaks unchanged and/or consistent (some are documented in the bug tracker, but dismissed). So far the devs do not consider them as a significant problem for their primary intended audience and environment (Windows-only schools and uni's in Russia)...

AFAIK, UTF-16/32 is seldom used in source files. Also, advanced programming editors supporting them are usually smart enough to distinguish them from UTF-8 regardless of BOM presence in the latter. Do you have any real issue with superfluous BOM in your environment?

P.S. I wonder why UTF-8 designers missed this Unicode recognition issue from single-byte encodings? They could try to officially define a custom magic sequence for UTF-8 as well (e.g. U+FEFFFFFE or something smarter). Although, fundamentally it's an OS and/or file-system design flaw -- there must be a standard and reliable file-format agnostic aparatus and method for setting/getting encoding on any file without guessing or hacky numbers.

**Reply by Sturwandan** [reply]
> I wonder why UTF-8 designers missed this Unicode recognition issue from single-byte encodings?

They did not, actually. The UTF-8 encoding is unique in multiple ways. First of all, as I mentioned before, it's fully compatible with ASCII, which is a must for every modern encoding, used to exchange data. Any starting sequence in the file itself would violate ASCII-compatibility, be it bogus BOM, or aforementioned U+FEFFFFFE.

But what you miss, is the fact that UTF-8 is easy to tell from legacy 8-bit encodings **without** any headers. Take a file and decode it as UTF-8. If there are no invalid sequences, then it actually is UTF-8. It's statistically impossible, that text in a legacy encoding would decode as UTF-8, because of the format having just enough redundancy.

In fact, UTF-8 is the most cleverly designed character encoding I have even seen.

> Do you have any real issue with superfluous BOM in your environment?

There are multiple issues. While they are not blockers, they are really annoying. For example BOM would break she-bang scripts, as in scripts which start with ``#!`` in the very beginning of file. Unix systems would recognize such files as executables, which are to be opened with a specific interpreter. For example ``#!/bin/bash`` or ``#!/bin/sh`` or ``#!/usr/bin/env perl`` are quite common. Of course, if bogus BOM gets between the # and beginning of file, the file won't work as executable anymore. 

BOM would make many compilers and interpreters hiccup, including aforementioned PascalABC (not .NET). If files with BOM are concatenated, BOM will end up in the middle of the file, which as you might guess is not nice.

> Since win-1251 single-byte encoding is still widely used in russian-oriented environment

Did you forget about CP866? It's even more common in Pascal files, if they were made under DOS with Russian strings. Can PascalABC.net tell CP866 from CP1251 automatically?

> statistical analysis of file content is not a reliable method to recognize proper encoding when you might have very few russian chars inside a source code.

Statistical analysis is not necessary if you want to tell UTF-8 from legacy encodings. See above. The method of trying to decode the file as UTF-8 and checking if everything decoded correctly is very reliable. Well-formed UTF-8 is not going to coincide with CP866 or CP1251, even if there are only several non-ASCII characters. 

> So, in international practice (at least from 80s) a 7-bit ASCII charset has almost always been just a common subset of other popular 8-bit encodings on PCs. That was long before any universal and internationally accepted multibyte standards such as MBCS/UCS/UTF became a reality and saw a wide adoption.

That's why BOM is unacceptable. It breaks the compatibility with 7-bit ASCII. Pure UTF-8 is fully compatible with ASCII, but BOM is not.
